benchmark-scripts/src/functions.py

```diff
@@ -219,54 +290,101 @@ def import_into_weaviate(client, efConstruction, maxConnections, benchmark_file)
+                        import_data_slice_to_weaviate,
+                        weaviate_url=weaviate_url,
+                        batch_size=benchmark_import_batch_size,
+                        vectors=data_to_import[start_indexes[i]:end_indexes[i]],
+                        subprocess_number=i,
+                        data_start_index=start_indexes[i]
+                    )
+                )
+            for f in as_completed(results):
+                try:
+                    f.result()
+                except Exception:
+                    loguru.logger.exception(
+                        'Something went wrong!'
+                    )
+                    import_failed = True
+    
+    end_time = time.monotonic()
+    if import_failed:
+        loguru.logger.error('Some import sub-processes failed! Check logs!')
+        exit(1)
+    
+    loguru.logger.info('done importing ' + str(c) + ' objects in ' + str(end_time - start_time) + ' seconds')
 
     return import_time
 
 
-def run_the_benchmarks(weaviate_url, CPUs, efConstruction_array, maxConnections_array, ef_array, benchmark_file_array):
+def run_the_benchmarks(
+        weaviate_url: str,
+        CPUs: int,
+        efConstruction_array: list,
+        maxConnections_array: list,
+        ef_array: list,
+        benchmark_file_array: list,
+    ):
     '''Runs the actual benchmark.
        Results are stored in a JSON file'''
 
     # Connect to Weaviate Weaviate
     try:
-        client = weaviate.Client(weaviate_url, timeout_config=(5, 60))
-    except:
+        # if weaviate is running in the same docker-compose.yml then this function is going to
+        # create a Client faster than Weaviate is ready, so we sleep 10 seconds
+        time.sleep(10)
+        client = Client(weaviate_url, timeout_config=(5, 60))
+    except Exception:
         print('Error, can\'t connect to Weaviate, is it running?')
-        exit(1)
+        print('Retrying to connect in 30 seconds.')
+        time.sleep(30)
+        try:
+            client = Client(weaviate_url, timeout_config=(5, 60))
+        except Exception:
+            print('Error, can\'t connect to Weaviate, is it running? Exiting ...')
+            exit(1)
 
     client.batch.configure(
         timeout_retries=10,
     )
 
-    # itterate over settings
+    # iterate over settings
     for benchmark_file in benchmark_file_array:
         for efConstruction in efConstruction_array:
             for maxConnections in maxConnections_array:
                
                 # import data
-                import_time = import_into_weaviate(client, efConstruction, maxConnections, benchmark_file)
+                import_time = import_into_weaviate(
+                    client=client,
+                    efConstruction=efConstruction,
+                    maxConnections=maxConnections,
+                    benchmark_file=benchmark_file,
+                    weaviate_url=weaviate_url,
+                    nr_cores=int(CPUs/2),
```

Comment:
what's the idea behind using `CPUs/2` instead of `CPUs` here?

Reply:
So, `CPUs` are basically threads, but processes need to be on per core, and since each core has 2 threads/CPUs it means we need half of the `CPUs` to have one process per core.